samples/containers/poetry/src/main/java/com/squareup/sample/poetry/PoemsBrowserWorkflow.kt

```diff
@@ -2,54 +2,219 @@ package com.squareup.sample.poetry
+import com.squareup.sample.poetry.PoemsBrowserWorkflow.State
+import com.squareup.sample.poetry.PoemsBrowserWorkflow.State.ComplexCall
+import com.squareup.sample.poetry.PoemsBrowserWorkflow.State.Initializing
+import com.squareup.sample.poetry.PoemsBrowserWorkflow.State.NoSelection
+import com.squareup.sample.poetry.PoemsBrowserWorkflow.State.Selected
 import com.squareup.sample.poetry.model.Poem
+import com.squareup.sample.poetry.views.BlankScreen
+import com.squareup.sample.poetry.views.LoaderSpinner
+import com.squareup.sample.poetry.views.MayBeLoadingScreen
 import com.squareup.workflow1.Snapshot
 import com.squareup.workflow1.StatefulWorkflow
+import com.squareup.workflow1.Worker
 import com.squareup.workflow1.action
 import com.squareup.workflow1.parse
+import com.squareup.workflow1.runningWorker
 import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
 import com.squareup.workflow1.ui.backstack.BackStackScreen
-
-typealias SelectedPoem = Int
+import kotlinx.coroutines.delay
 
 object PoemsBrowserWorkflow :
-  StatefulWorkflow<List<Poem>, SelectedPoem, Nothing, OverviewDetailScreen>() {
+  StatefulWorkflow<Config, State, Nothing, MayBeLoadingScreen>() {
+
+  data class Config(
+    val poems: List<Poem>,
+    val simulatedPerfConfig: SimulatedPerfConfig = SimulatedPerfConfig.NO_SIMULATED_PERF
+  )
+
+  sealed class State {
+    var poemLoading = false
+
+    // N.B. This state is a smell. We include it as possible in order to be able to config similar
+    // to real-world usage. It should be folded into [NoSelection] at the very least.
+    object Initializing : State()
+    data class ComplexCall(
+      val payload: Int
+    ) : State()
+
+    object NoSelection : State()
+    data class Selected(val poemIndex: Int) : State()
+  }
 
   override fun initialState(
-    props: List<Poem>,
+    props: Config,
     snapshot: Snapshot?
-  ): SelectedPoem {
-    return snapshot?.bytes?.parse { source -> source.readInt() }
-      ?: NO_POEM_SELECTED
+  ): State {
+    return snapshot?.bytes?.parse { source ->
+      if (props.simulatedPerfConfig.isComplex) {
+        ComplexCall(source.readInt())
+      } else {
+        val selected = source.readInt()
+        if (selected != NO_POEM_SELECTED) {
+          Selected(selected)
+        } else {
+          NoSelection
+        }
+      }
+    } ?: if (props.simulatedPerfConfig.useInitializingState) Initializing else NoSelection
   }
 
   @OptIn(WorkflowUiExperimentalApi::class)
   override fun render(
-    renderProps: List<Poem>,
-    renderState: SelectedPoem,
+    renderProps: Config,
+    renderState: State,
     context: RenderContext
-  ): OverviewDetailScreen {
-    val poems: OverviewDetailScreen =
-      context.renderChild(PoemListWorkflow, renderProps) { selected -> choosePoem(selected) }
-        .copy(selection = renderState)
-        .let { OverviewDetailScreen(BackStackScreen(it)) }
-
-    return if (renderState == NO_POEM_SELECTED) {
-      poems
+  ): MayBeLoadingScreen {
```

Comment:
The loading screen is another example of my concern about this sample becoming too confusing to be useful as such. I'd like it to stay focused on BackStackScreen and OverviewDetailScreen.

Reply:
It should be possible to add loading to the benchmark version w/o requiring it to leak into the "public API" of the sample. E.g. take a gatekeeper approach:

- the benchmark versions of this workflow injects a relay that controls whether the loading screen is showing
- an outer-most workflow used only by the benchmark implementation monitors that relay, and wraps its children's renderings in a `MayBeLoadingScreen`.

Reply:
This is where I thought you were suggesting the injected loading trigger @rjrjr ? But now it seems like you're having second thoughts about that - should we move it back to the rendering?

Reply:
I'm saying that the Gatekeeper approach was suggested as a way to keep all knowledge of the loading stuff out of the non-benchmark workflow. Loading gets implemented as a wrapper around `PoemWorkflow`, without its knowledge.

* The root UI of the Benchmark version uses something like `MaybeLoading<PoemWorkflow>`, or `MaybeLoading<*>` as its `RenderingT`. It runs a `PoemWorkflow` as a child.
* `PerformancePoemWorkflow` is in on the joke. It injects the `Loader` service interface I just suggested to make it known when it enters or exits a loading state.

Reply:
Ack - ok that makes a lot of sense. I will add a note saying that we should wrap it in an interface but that isn't shown here for the sake of simplicity.

Reply:
- [ ] Add note about smelliness of shared mutable state.